#484 Endpoint for downloading logfiles locally

IvanObreshkov opened the pull request:
Created an endpoint `GET /logs` for downloading the `logfile.log ` from our prod server when deployed.

Resolves #484   

## How to verify:
- [ ] Change the `IS_OFFLINE` value in your .env file to `false`
- [ ] Go to the `auth.py` middleware and add the endpoint to the  BYPASSED_ENDPOINTS dict:
![image](https://github.com/AUBGTheHUB/monolith/assets/38375079/1b6be7e9-f1f3-4b27-a2c5-8bc47986f74f)

- [ ] Open your browser and head to [/v2/logs](http://localhost:6969/v2/logs).

The file should appear in the folder where your Downloads are stored by default. If no text or page appears on your screen this is completely normal as we are returning a `FileResponse`. If the file is empty this is normal as well.
I encourage you to come up with other edge cases, if possible, that are not handled in my code.

asynchroza commented:
I don't have the time to review it now but from what I see in the description, this endpoint is publicly accessible and it shouldn't be.

asynchroza commented on packages/py-api/py_api/controllers/getting_logs_controller.py:
```diff
@@ -0,0 +1,44 @@
+                },
+            )
+
+        except Exception as e:
```
This is not needed. This is handled by the exception handler.

IvanObreshkov replied:
Yes. None of the methods I use will raise an Exception so I don't need it.

asynchroza commented on packages/py-api/py_api/controllers/getting_logs_controller.py:
```diff
@@ -0,0 +1,44 @@
+from py_api.environment import IS_OFFLINE
+from starlette.responses import FileResponse, JSONResponse
+
+router = APIRouter()
```
Why is there a router in the controller?

asynchroza commented on packages/py-api/py_api/controllers/getting_logs_controller.py:
```diff
@@ -0,0 +1,44 @@
+
+from fastapi import APIRouter
+from py_api.environment import IS_OFFLINE
+from starlette.responses import FileResponse, JSONResponse
```
We usually use the JSONResponse from `fastapi.responses` not starlette

IvanObreshkov replied:
Fixed

asynchroza commented on packages/py-api/py_api/controllers/getting_logs_controller.py:
```diff
@@ -0,0 +1,44 @@
+from starlette.responses import FileResponse, JSONResponse
+
+router = APIRouter()
+LOGFILE_PATH: Final = f"{pathlib.Path(__file__).parent.resolve().parent}/shared/logfile.log"
```
Let's import Path from path lib instead of having it like this

asynchroza replied:
pathlib.Path(__file__).parent.resolve().parent
                                                               ^^^^^^^
                                                         Are you sure you need this?

asynchroza replied:
Oh yes, you do. My bad. You're in a child directory.

IvanObreshkov replied:
> Let's import Path from path lib instead of having it like this

Fixed

asynchroza commented on packages/py-api/py_api/controllers/getting_logs_controller.py:
```diff
@@ -0,0 +1,44 @@
+LOGFILE_PATH: Final = f"{pathlib.Path(__file__).parent.resolve().parent}/shared/logfile.log"
+
+
+class GettingLogsController:
```
rename to LogsController

asynchroza commented on packages/py-api/py_api/controllers/getting_logs_controller.py:
```diff
@@ -0,0 +1,44 @@
+
+class GettingLogsController:
+    @classmethod
+    def return_files(cls) -> JSONResponse | FileResponse:
```
Why do you have two methods `get_log_file` and `get_files`. 
You should learn to call the functions by what they are actually doing. 

I expect to load a handler for a file or get a blob/bytes array when calling `get_log_file`, not a JSONResponse or a FileResponse.

asynchroza replied:
No need for two separate methods here. 
Combine them and drop the try catch clause.
Check whether the file exists and that we're in the correct environment in the beginning of the method. 

Return appropriate response, no need for handling other exceptions than being in the incorrect environment and logs not being present.

asynchroza commented on packages/py-api/py_api/controllers/getting_logs_controller.py:
```diff
@@ -0,0 +1,44 @@
+                return JSONResponse(content={"error": "The logs folder is empty"}, status_code=404)
+
+            return FileResponse(
+                LOGFILE_PATH, headers={
```
Will this error out if the file is not present. If so, is the exception descriptive enough? If that's the case, drop handling for inexistent file. This will be handled by the error handler.

IvanObreshkov replied:
It will not raise an exception

asynchroza commented on packages/py-api/py_api/routes/getting_logs_routes.py:
```diff
@@ -0,0 +1,10 @@
+
+@router.get("/logs")
+async def get_logs() -> JSONResponse:
+    return c.return_files()
```
c.get_logs or c.get_log_file 

whichever you prefer more

asynchroza commented on packages/py-api/py_api/routes/getting_logs_routes.py:
```diff
@@ -0,0 +1,10 @@
+from fastapi import APIRouter
+from py_api.controllers.getting_logs_controller import GettingLogsController as c
+from starlette.responses import JSONResponse
```
Same here, we don't use starlette's responses

asynchroza commented on packages/py-api/py_api/routes/getting_logs_routes.py:
```diff
@@ -0,0 +1,10 @@
+from fastapi import APIRouter
```
File should be renamed to logs_routes

IvanObreshkov replied:
Done

asynchroza commented on packages/py-api/py_api/routes/routes.py:
```diff
@@ -2,6 +2,7 @@
 
 from fastapi import APIRouter, FastAPI
 from py_api.routes.feature_switches_routes import router as feature_switches_router
+from py_api.routes.getting_logs_routes import router as get_logs_router
```
logs_router

asynchroza commented on packages/py-api/py_api/controllers/getting_logs_controller.py:
```diff
@@ -0,0 +1,44 @@
+import pathlib
```
file should be renamed to logs_controller

IvanObreshkov replied:
Done

asynchroza commented:
> I don't have the time to review it now but from what I see in the description, this endpoint is publicly accessible and it shouldn't be.

It seems that I've read the description incorrectly. Authentication looks good.

asynchroza commented:
I couple of comments, otherwise it works as expected.

asynchroza commented on packages/py-api/py_api/controllers/getting_logs_controller.py:
```diff
@@ -0,0 +1,44 @@
+
+        if IS_OFFLINE:
+            content = {
+                "message": "You should be in a prod environment. Log files are not present in a local one.",
```
Let's change it to -> "Log files are not available in local environments."

IvanObreshkov commented:
> I couple of comments, otherwise it works as expected.

I believe all comments are resolved. You can review it again.

asynchroza commented on packages/py-api/py_api/controllers/logs_controller.py:
```diff
@@ -0,0 +1,30 @@
+
+class LogsController:
+    @classmethod
+    def get_log_file(cls) -> JSONResponse | FileResponse:
```
Since you don't call any class functions, drop the `@classmethod` decorator here.

asynchroza replied:
ooor put encapsulate the LOGFILE_PATH variable in the class itself and accessing by using `cos.LOGFILE_PATH`

asynchroza replied:
I would prefer the second but it's up to you.

IvanObreshkov replied:
@asynchroza Yes but in a separate issue.

asynchroza replied:
@IvanObreshkov then create one and I will approve this PR.

asynchroza commented on packages/py-api/py_api/controllers/logs_controller.py:
```diff
@@ -0,0 +1,30 @@
+            return JSONResponse(content=content, status_code=status_code)
+
+        if not exists(LOGFILE_PATH):
+            return JSONResponse(content={"error": "The logs folder is empty"}, status_code=404)
```
change error to message

IvanObreshkov replied:
Done

asynchroza commented on packages/py-api/py_api/controllers/logs_controller.py:
```diff
@@ -0,0 +1,30 @@
+            return JSONResponse(content=content, status_code=status_code)
+
+        if not exists(cls.LOGFILE_PATH):
+            return JSONResponse(content={"message": "The logs folder is empty"}, status_code=404)
```
This message is not true per se. The folder includes the README and it might include backup logs but for some reason miss the primary one. Change it `Log file is not existing`

asynchroza commented on packages/py-api/py_api/controllers/logs_controller.py:
```diff
@@ -0,0 +1,30 @@
+
+        if IS_OFFLINE:
+            content = {
+                "message": "Log files are not available in local environments.",
```
remove the dot at the end 😆 I'm sorry for making you do this but let's stick to doing the same for all of our response. Have dots only if the messages are more than one sentence.

asynchroza commented:
@IvanObreshkov 

A couple more nits and it should be good to go.
Please, don't comment `Done`. Just ping me once when you've addressed all comments.

IvanObreshkov commented:
> @IvanObreshkov
> 
> A couple more nits and it should be good to go. Please, don't comment `Done`. Just ping me once when you've addressed all comments.

@IvanObreshkov we should be good to go. Also what are your thoughts on https://github.com/AUBGTheHUB/monolith/pull/492#discussion_r1307348531

asynchroza reviewed (approved):
Thank you for working on this one! Nice job!
